### PR TITLE
Custom marshallers v2 usage

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Convert some custom type marshallers to the v2 design.

Convert as many of the marshallers as possible to their correct representation with the currently implemented feature set.

This PR does not convert marshallers that can be represented with the new design but need more features to be accurately represented.

Viewing this PR with no whitespace diffs is highly recommended, as the changes are very mechanical but the whitespace changes make them look much worse.

### CODE REVIEW

File: src/libraries/System.Diagnostics.EventLog/src/System/Diagnostics/Reader/UnsafeNativeMethods.cs [hunk 1]
```diff
@@ -703,37 +703,44 @@ internal struct EvtStringVariant
...
-
-                public Native(EvtStringVariant managed)
-                {
-                    StringVal = Marshal.StringToCoTaskMemUni(managed.StringVal);
-                    Count = managed.Count;
-                    Type = managed.Type;
-                }
+                public static Native ConvertToUnmanaged(EvtStringVariant managed) => new(managed);
+                public static EvtStringVariant ConvertToManaged(Native native) => native.ToManaged();
+                public static void FreeNative(Native native) => native.FreeNative();
```
Comment (elinor-fung) [inline src/libraries/System.Diagnostics.EventLog/src/System/Diagnostics/Reader/UnsafeNativeMethods.cs:711]:
```suggestion
                public static void Free(Native native) => native.FreeNative();
```

File: src/libraries/System.Drawing.Common/src/Interop/Windows/Interop.Gdi32.cs [hunk 2]
```diff
@@ -188,29 +188,35 @@ internal sealed class DOCINFO
...
+            [ManagedToUnmanagedMarshallers(typeof(DOCINFO), InMarshaller = typeof(Marshaller))]
+            public static class Marshaller
             {
-                internal int cbSize;
-                internal IntPtr lpszDocName;
-                internal IntPtr lpszOutput;
-                internal IntPtr lpszDatatype;
-                internal int fwType;
+                public static Native ConvertToUnmanaged(DOCINFO managed) => new(managed);
+                public static void FreeNative(Native native) => native.FreeNative();
```
Comment (elinor-fung) [inline src/libraries/System.Drawing.Common/src/Interop/Windows/Interop.Gdi32.cs:195]:
```suggestion
                public static void Free(Native native) => native.FreeNative();
```

File: src/libraries/System.Runtime.InteropServices/tests/LibraryImportGenerator.Tests/LibraryImportGenerator.Tests.csproj [hunk 3]
```diff
@@ -3,6 +3,7 @@
     <TargetFramework>$(NetCoreAppCurrent)</TargetFramework>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <EnableLibraryImportGenerator>true</EnableLibraryImportGenerator>
+    <InculdeLibraryImportGeneratorSources>false</InculdeLibraryImportGeneratorSources>
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/tests/LibraryImportGenerator.Tests/LibraryImportGenerator.Tests.csproj:6]:
```suggestion
    <IncludeLibraryImportGeneratorSources>false</IncludeLibraryImportGeneratorSources>
```

File: src/libraries/System.Drawing.Common/src/Interop/Windows/Interop.Gdi32.cs [hunk 4]
```diff
@@ -188,29 +188,35 @@ internal sealed class DOCINFO
...
+            [ManagedToUnmanagedMarshallers(typeof(DOCINFO), InMarshaller = typeof(Marshaller))]
+            public static class Marshaller
             {
-                internal int cbSize;
-                internal IntPtr lpszDocName;
-                internal IntPtr lpszOutput;
-                internal IntPtr lpszDatatype;
-                internal int fwType;
+                public static Native ConvertToUnmanaged(DOCINFO managed) => new(managed);
+                public static void Free(Native native) => native.FreeNative();
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Drawing.Common/src/Interop/Windows/Interop.Gdi32.cs:195]: Should we be concerned with the size of the copy in this case? This doesn't seem too bad, but if a native type is large enough, no?

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Drawing.Common/src/Interop/Windows/Interop.Gdi32.cs:195]: Is a `Free(ref Native native)` better?

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Review by elinor-fung — APPROVED** [review]

File: src/libraries/System.Drawing.Common/src/Interop/Windows/Interop.Gdi32.cs (hunk 4 above)
Comment (jkoritzinsky) [reply src/libraries/System.Drawing.Common/src/Interop/Windows/Interop.Gdi32.cs:195]: That's an interesting point and definitely worth considering. Maybe we can optionally allow this to be a `ref` parameter (I don't know if I'd want to require it to be a `ref`, but we could also require it).

Comment (jkoritzinsky) [reply src/libraries/System.Drawing.Common/src/Interop/Windows/Interop.Gdi32.cs:195]: We can talk about this in API review on Tuesday

**Pull request merged**